Mathlib/Geometry/Euclidean/Sphere/Tangent.lean

```diff
@@ -0,0 +1,405 @@
+
+lemma mem_commonIntTangents_iff {as : AffineSubspace ℝ P} {s₁ s₂ : Sphere P} :
+    as ∈ s₁.commonIntTangents s₂ ↔
+      as ∈ s₁.commonTangents s₂ ∧ ∃ p ∈ as, Wbtw ℝ s₁.center p s₂.center :=
+  Iff.rfl
+
+lemma mem_commonExtTangents_iff {as : AffineSubspace ℝ P} {s₁ s₂ : Sphere P} :
+    as ∈ s₁.commonExtTangents s₂ ↔
+      as ∈ s₁.commonTangents s₂ ∧ ∀ p ∈ as, ¬Sbtw ℝ s₁.center p s₂.center :=
+  Iff.rfl
+
+@[simp] lemma union_commonIntTangents_commonExtTangents (s₁ s₂ : Sphere P) :
+    s₁.commonIntTangents s₂ ∪ s₁.commonExtTangents s₂ = s₁.commonTangents s₂ := by
+  ext as
+  rw [Set.mem_union, mem_commonIntTangents_iff, mem_commonExtTangents_iff, ← and_or_left,
+    and_iff_left_iff_imp]
+  rintro -
+  by_cases h : ∃ p ∈ as, Wbtw ℝ s₁.center p s₂.center
+  · exact .inl h
+  · refine .inr ?_
+    simp_rw [not_exists, not_and] at h
+    rintro p hp
+    exact mt Sbtw.wbtw (h p hp)
+
+/-- The spheres `s₁` and `s₂` are externally tangent at the point `p`. -/
+structure IsExtTangentAt (s₁ s₂ : Sphere P) (p : P) : Prop where
+  mem_left : p ∈ s₁
+  mem_right : p ∈ s₂
+  wbtw : Wbtw ℝ s₁.center p s₂.center
+
+lemma IsExtTangentAt.symm {s₁ s₂ : Sphere P} {p : P} (h : s₁.IsExtTangentAt s₂ p) :
+    s₂.IsExtTangentAt s₁ p where
+  mem_left := h.mem_right
+  mem_right := h.mem_left
+  wbtw := h.wbtw.symm
+
+lemma isExtTangentAt_comm {s₁ s₂ : Sphere P} {p : P} :
+    s₁.IsExtTangentAt s₂ p ↔ s₂.IsExtTangentAt s₁ p :=
+  ⟨IsExtTangentAt.symm, IsExtTangentAt.symm⟩
+
+lemma isExtTangentAt_center_iff {s₁ s₂ : Sphere P} :
+    s₁.IsExtTangentAt s₂ s₁.center ↔ s₁.radius = 0 ∧ s₁.center ∈ s₂ := by
+  refine ⟨?_, ?_⟩
+  · rintro ⟨h₁, h₂, -⟩
+    rw [center_mem_iff] at h₁
+    exact ⟨h₁, h₂⟩
+  · rintro ⟨hr, hc⟩
+    refine ⟨?_, hc, ?_⟩
+    · rw [center_mem_iff, hr]
+    · simp
+
+/-- The sphere `s₁` is internally tangent to the sphere `s₂` at the point `p` (that is, `s₁` lies
+inside `s₂` and is tangent to it at that point). This includes the degenerate case where the
+spheres are the same. -/
+structure IsIntTangentAt (s₁ s₂ : Sphere P) (p : P) : Prop where
+  mem_left : p ∈ s₁
+  mem_right : p ∈ s₂
+  wbtw : Wbtw ℝ s₂.center s₁.center p
+
+lemma isIntTangentAt_center_iff {s₁ s₂ : Sphere P} :
+    s₁.IsIntTangentAt s₂ s₁.center ↔ s₁.radius = 0 ∧ s₁.center ∈ s₂ := by
+  refine ⟨?_, ?_⟩
+  · rintro ⟨h₁, h₂, -⟩
+    rw [center_mem_iff] at h₁
+    exact ⟨h₁, h₂⟩
+  · rintro ⟨hr, hc⟩
+    refine ⟨?_, hc, ?_⟩
+    · rw [center_mem_iff, hr]
+    · simp
+
+@[simp] lemma isIntTangentAt_self_iff_mem {s : Sphere P} {p : P} :
+    s.IsIntTangentAt s p ↔ p ∈ s :=
+  ⟨fun ⟨h, _, _⟩ ↦ h, fun h ↦ ⟨h, h, by simp⟩⟩
+
+/-- The spheres `s₁` and `s₂` are externally tangent at some point. -/
+def IsExtTangent (s₁ s₂ : Sphere P) : Prop :=
+  ∃ p, s₁.IsExtTangentAt s₂ p
+
+/-- The spheres `s₁` and `s₂` are internally tangent at some point. -/
+def IsIntTangent (s₁ s₂ : Sphere P) : Prop :=
+  ∃ p, s₁.IsIntTangentAt s₂ p
```

Comment:
Could you please add `.symm` and `_comm` lemmas for these, that show that this is a symmetric relation.

Reply:
Added for `IsExtTangent`. `IsIntTangent` isn't symmetric, I've clarified the doc string there.